[초간단 계산기 구현] 이지현 과제 제출합니다.

izzy80 opened the pull request.

izzy80 committed:
feat:Calculator Class 구현

izzy80 committed:
chore: gradle에 assertj-core 의존성 추가

izzy80 committed:
test: Calculator 테스트 코드 작성

Choi-JJunho commented on src/main/java/Calculator.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * 기능 요구사항
+ * -인자 2개를 받아 사칙연산을 할 수 있는 계산기를 구현한다
+ * -사칙연산과 매칭되는 4개의 메서드를 제공한다.
+ * -계산된 결과는 정수를 반환한다.
```
꼼꼼한 요구사항 작성 좋네요~! 👍
요구사항을 모아두는 README.md 문서를 작성해보는건 어떨까요?

izzy80 replied:
좋은 생각인 것 같습니다!

Choi-JJunho commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    @DisplayName("assertThatThrownBy:divide 메서드는 0으로 나누면 ArithmeticException을 던진다")
```
0으로 나누는 예외상황까지~! 꼼꼼한 테스트 좋습니다 👍

Choi-JJunho commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+    @Test
+    @DisplayName("assertEquals:add_두수의_합을_반환")
+    void add_두수의_합을_반환() {
+        final var a = 3;
+        final var b = 4;
+        final var actual = 7;
```
var를 사용하셨군요!
지현님은 어떨 때 var 사용을 선호하시나요?
혹은 var를 선호하는 이유가 있으실까요?

izzy80 replied:
사실 이번에 java-learning-test 공부하면서 처음 사용해봅니다! 기존에는 타입에 맞춰서 쓰곤합니다. 실제로 실무에서 var을 많이 쓰는지 궁금합니다.

Choi-JJunho replied:
var의 사용은 취향차이인 것 같아요 ㅎㅎ
저는 개인적으로 추가로 활용하지 않고 반환하는 값을 변수로 잠시 빼둘 때는 var로 선언하고, 나머지는 타입에 맞춰 쓰는 편인 것 같아요!

Choi-JJunho commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+    Calculator calculator = new Calculator();
+
+    @Test
+    @DisplayName("assertEquals:add_두수의_합을_반환")
```
저는 개인적으로 메소드를 한글로 작성하면 충분히 의도를 전달할 수 있다고 생각해요!
DisplayName으로 동일한 테스트명을 한번 더 선언해주신 이유가 있을까요?

izzy80 replied:
기존에 메소드명을 영어로 쓸려고 했는데 복잡한 것 같아서 한글로 쓰다보니 테스트명이랑 메소드명에 동일한 내용을 기술했습니다.

Choi-JJunho commented on src/main/java/Calculator.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public class Calculator {
+
+    int add(int num1, int num2) {
+        return num1 + num2;
+    }
+
+    int subtract(int num1, int num2) {
+        return num1 - num2;
+    }
+
+    int multiply(int num1, int num2) {
+        return num1 * num2;
+    }
+
+    int divide(int num1, int num2) {
+        return num1 / num2;
+    }
```
사칙연산을 계산기라는 객체의 메소드로 만드는것 과 static 메소드로 만드는 것 각각 어떤 장/단점이 있을까요?

izzy80 replied:
## 객체의 메소드
객체의 상태를 조회하거나 변경에 사용
인스턴스 메서드는 객체의 상태와 행동을 하나의 단위로 묶어, 캡슐화와 정보 은닉을 강화
### 장점
1. 객체 지향 프로그래밍의 원칙에 부합
객체의 상태와 행동을 하나의 단위로 묶어, 캡슐화와 정보 은닉을 강화
2. 다형성을 지원
상속받은 클래스에서 오버라이딩할 수 있으므로, 동일한 인터페이스 아래에서 다양한 구현을 제공
### 단점
1. 객체의 생성이 필요하다는 점에서 오버헤드가 발생
각 메서드 호출 전에 객체를 생성해야 하므로, 불필요한 객체 생성으로 인한 메모리 사용량 증가가 우려

## static 메소드
유틸리티 함수, 헬퍼 함수, 팩토리 메서드 등에 사용
객체의 상태와 무관하게 일반적인 연산을 수행
### 장점
1. 사용의 편리성
객체를 생성하지 않고도 메서드를 바로 호출 가능
초기화나 설정 작업을 수행할 때 유용
### 단점
1. 객체 지향 프로그래밍의 원칙에 어긋남
객체의 상태와 행동을 분리시키므로, 객체 지향 설계의 목적인 상태와 행동의 캡슐화를 약화
2. 테스트와 유지보수 측면에서 제약
오버라이딩이 불가능하므로, 테스트 시에 다른 구현으로 대체하기 어렵다

---
덕분에 어렴풋이 알고만 있었는데 자바 공부 다시 했습니다ㅎㅎ

Choi-JJunho replied:
객체의 메소드에 다음과 같은 단점을 작성해주셨네요!

> 각 메서드 호출 전에 객체를 생성해야 하므로, 불필요한 객체 생성으로 인한 메모리 사용량 증가가 우려

JVM은 이런 불필요한 객체 생성을 어떻게 해소하고있을까요? (Hint: GC)

izzy80 replied:
`자바 가비지 컬렉션이 자바 가상 머신의 힙 영역에서 사용되지 않는 객체들을 자동으로 찾아내고, 이를 제거하여 메모리를 회수합니다. 이를 통해 개발자가 직접 메모리를 관리하지 않아도 되게 하여, 메모리 누수와 같은 문제를 방지할 수 있습니다. `

찾아보다가 그러면 GC가 알아서 해결해주면 단점이 아니지 않을까? 했는데 결국에는 객체 생성 자체에도 비용이 들기 때문에 객체를 아예 만들지 않는 것보다 효율적일 수 없다고 하네요. 가능한 객체 생성을 줄이고 재사용하는 방향으로 설계해야겠어요!

Choi-JJunho commented on src/test/java/CalculatorTest.java:
```diff
@@ -0,0 +1,81 @@
+        }).isInstanceOf(ArithmeticException.class)
+                .hasMessage("/ by zero");
+    }
+
+
+
+}
```
(알쓸신잡)
![image](https://github.com/user-attachments/assets/042a3550-9cbc-43be-b636-e5c483de494c)

Github를 보다보면 이런 빨간색 경고 아이콘이 뜨곤 하는데요! 무슨 문제일까요~?
[도움이 될만한 글](https://velog.io/@junho5336/No-newline-at-end-of-file) 하나를 가져와봤습니다!

izzy80 replied:
주노 글이네요!! 경고는 파일의 가장 마지막 줄에는 개행이 있어야한다는 경고네요. 새로운 정보 얻어갑니다😁

Choi-JJunho reviewed (commented):
첫 미션 제출 너무 고생하셨습니다~!
꼼꼼한 테스트 작성이 매우 인상적이였어요 👍 

몇가지 궁금한점에 대해 코멘트를 남겼으니 확인 부탁드릴게요!

또한 몇가지 확인해주실 내용이 있어 코멘트로 남겨봅니다~!
---

### 계산기 - 단위테스트 의 모든 단계를 수행해주세요!

![image](https://github.com/user-attachments/assets/69285e88-8b22-4309-b8fa-cdc065e542a4)

모든 단계를 수행하여 commit을 올려주세요~!
진행 과정에서 어렵거나 잘 모르는 부분이 있다면 편하게 질문해주세요!

izzy80 committed:
docs:단계별 요구사항 README 파일

izzy80 committed:
chore: 클래스 내 불필요한 주석 제거

izzy80 committed:
feat: StringCalculator Class 구현

izzy80 committed:
chore: StringCalculator Class print문 주석처리

izzy80 committed:
feat: 문자열 계산기 substring idx 수정

izzy80 committed:
test:  StringCalculator 테스트 코드 작성

izzy80 committed:
test: StringCalculator 테스트 코드 AssertJ로 리팩터링

Choi-JJunho commented on README.md:
```diff
@@ -1,3 +1,28 @@
-
 계산기 미션 저장소
+
+## 1단계-초간단 계산기 구현
```
꼼꼼한 README 좋습니다 👍

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,32 @@
+public class StringCalculator {
+    int sum(String str){
+//        System.out.println(str);
```
사용하지 않는 주석은 제거해주셔도 좋을 것 같아요!

izzy80 replied:
넵

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,32 @@
+        if(str.contains("-")){//음수이면
+            throw new RuntimeException();
+        }
+        if(str.matches(".*[가-힣ㄱ-ㅎㅏ-ㅣa-zA-Z].*")){//숫자 이외의 값이 들어온다면
```
숫자 이외의 값에 대한 정규식이 다소 어렵게 작성되어있는 것 같아요!
이 값들은 한자, 특수문자 등을 모두 포함하는 정규식인가요?

izzy80 replied:
숫자 이외의 상황이 들어오면 안 된다 -> 숫자만 허락할까? -> 구분자도 들어있는데? 
-> 그럼 숫자 이외의 값을 정규식에 넣자!
라는 흐름으로 구현했는데 그러다보니 좀 더 복잡한 예외상황을 고려하지 못해네요 😱

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,32 @@
+public class StringCalculator {
+    int sum(String str){
```
메소드가 너무 많은 책임을 지고 있는 것 같아요!
단일 책임 원칙을 한번 지켜보시죠!

현재 메소드가 하고있는 일을 정리하고 메소드로 분리해볼까요?

izzy80 replied:
이전)
sum 메소드
이후)
1. sum 메소드
2. splitString 메소드
- input값을 받아서 Integer 리스트로 반환하는 메소드
3. finalDelimiter
- 커스텀 구분자가 2개 이상일 경우 처리하기 위해 메소드 별개로 빼냄
4. parseIntOrThrow 메소드
- 음수가 있다면 예외처리
- 숫자 이외의 값이 들어온다면 예외처리

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,32 @@
+
+
+        int total = 0;
+        String del = "";//구분자
+        if(str.contains("//")){
+            del = str.substring(2,3);
+            str = str.substring(4);
+        }
```
del, str과 같은 축약된 변수명은 의도를 파악하기가 어려운 것 같아요!
보다 명시적인 이름으로 변경해볼까요?

izzy80 replied:
변수 이름 정하기는 항상 어려운 것 같네요 🫠 수정하도록 하겠습니다!

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,32 @@
+            del = "[,|:]";
+        }
+
+        String[] arr = str.split(del);
+        for(String s : arr){
+            total += Integer.parseInt(s);
+        }
```
여기도 마찬가지로 arr, s 와 같은 변수명은 의도를 파악하기가 어려운것 같아요~!

Choi-JJunho commented on src/test/java/StringCalculatorTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+
+    //Refactoring
```
이 주석은 어떤 의미일까요?

izzy80 replied:
Junit에서 assertJ로 리팩토링하라는 말이 있어서 중간에 주석을 적어놓았습니다!

Choi-JJunho commented on src/test/java/StringCalculatorTest.java:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    @Test
+    @DisplayName("\"/\" 입력 시 오류가 발생하지 않고 0이 출력된다")
```
예외테스트 좋네요 👍

Choi-JJunho reviewed (commented):
구현하시느라 고생하셨습니다~!
3단계까지 구현한 내용에 대해 코멘트 몇가지 남겼습니다
확인 부탁드릴게요~~

izzy80 committed:
refactor: StringCalculator 리팩토링
- 변수 이름을 명확하게 변경하여 가독성 향상
- 메소드 단일 책임 원칙(SRP)에 따라 기능 분리
- 불필요한 주석 제거로 코드 정리
- 구분자 예외처리

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,61 @@
+        return result.toString();
+    }
+
+    int parseIntOrThrow (String token){
```
예외처리에 대한 부분까지는 메소드명으로 드러낼 필요는 없어보이는데 어떻게 생각하시나요~?

izzy80 replied:
음 parseInt로 그대로 쓰면 메소드명이 겹쳐서 뒤에 에러처리에 대한 것까지 넣었는데 메소드명으로 권장되는 방식이 아니군요! 예외처리는 보통 함수에서 다 처리하는 것이라서 굳이 메소드명에 드러날 필요가 없어서 그런거겠죠?? 메소드명은 좀 더 기능에 의해서 서술해야겠네요. 

stringToInt()로 수정하도록 하겠습니다!

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,61 @@
+
+public class StringCalculator {
+    static final String DEFAULT_DELIMITER = ",|:";
+    static final String REGEX = "//(.*)\n(.*)";
```
무엇을 나타내는 REGEX일까요?
명시적인 네이밍으로 나타내볼 수 있을까요?

izzy80 replied:
넵!

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,61 @@
+    static final String REGEX = "//(.*)\n(.*)";
+
+    int sum(String input){
+        if(input == null || input.isEmpty()){
+            return 0;
+        }
```
이 상황에서는 예외 처리가 아닌 0을 반환한 이유가 있을까요?

izzy80 replied:
![image](https://github.com/user-attachments/assets/528bb534-8485-4c32-9561-cb5c247acc23)
요구사항에서 ""가 입력될 시 0으로 출력하라고 해서 sum(String input)에서 ""이 입력될 때 바로 0값을 return되게 만들었습니다. 

답변이 질문의 의도에 맞게 답이 되었나요?

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,61 @@
+            return 0;
+        }
+        List<Integer> numbers = splitString(input);
+        return numbers.stream().mapToInt(i->i).sum();
```
String을 사용하셨네요!
stream 내부에서 예외가 발생하면 어떤 일이 일어날까요?

예를들어 int 변환 중 에러가 발생하면 어떤 일이 일어날까요?

izzy80 replied:
int변환은 parseIntOrThrow에서 이루어지기 때문에 `return numbers.stream().mapToInt(i->i).sum();` 이 부분은 list에 있는 값을 더해서 합을 출력하기만 수행합니다. 

하지만 만약 stream 내부에서 Integer.parseInt()와 같은 문자열 → 숫자 변환이 이루어진다면, NumberFormatException 또는 RuntimeException가 발생해서 stream 함수가 제대로 동작이 되지 않고 중간에 에러를 뱉어낼 것으로 예상됩니다!

Choi-JJunho commented on src/main/java/StringCalculator.java:
```diff
@@ -0,0 +1,61 @@
+        return numbers.stream().mapToInt(i->i).sum();
+    }
+
+    List<Integer> splitString(String input) {
```
내부에서만 사용되는 메소드라면 private 접근제어자를 붙여보는건 어떨까요?

+) 지금처럼 접근제어자를 지정하지 않으면 메소드의 허용범위가 어디까지 가능해질까요?

izzy80 replied:
지정하지 않는다면 접근제어자가 default이기 때문에 같은 패키지에서만 사용될 수 있습니다. 만약 외부에서 해당 객체를 생성하게 될 경우, 메서드가 자동완성 목록에 노출될 수 있어 개발자에게 혼동을 줄 수 있습니다.

답변하다가 궁금해서 default면 test에서 어떻게 메소드를 이용하는거지? 다른 패키지 아닌가? 생각했는데 찾아보니 main과 test는 패키지가 아니라, 디렉토리 구조이자 빌드 단위라고 하네요. 
패키지는 java부터 시작되어서 결국에는 main의 java와 test의 java 패키지가 같은 패키지라는 것을 알게되었습니다!

Choi-JJunho reviewed (approved):
추가된 내용까지 고생하셨습니다😊
궁금한 내용 추가로 남겨봤으니 확인 부탁드려요!

izzy80 committed:
refactor: StringCalculator 리팩토링
- 접근 제어자(private) 명시
- 메서드명 변경 (splitString → splitInput, parseIntOrThrow → stringToInt)
- 상수명 수정 (REGEX → CUSTOM_REGEX)
